xmppserver/src/main/java/org/jivesoftware/openfire/container/PluginMonitor.java

```diff
@@ -446,16 +435,32 @@ public Integer call() throws Exception
          * @param pluginName the name of the plugin.
          * @param file       the JAR file
          * @param dir        the directory to extract the plugin to.
+         * @return A boolean indicating success.
          */
-        private void unzipPlugin( String pluginName, Path file, Path dir )
+        private boolean unzipPlugin( String pluginName, Path file, Path dir )
         {
             try ( ZipFile zipFile = new JarFile( file.toFile() ) )
             {
                 // Ensure that this JAR is a plugin.
                 if ( zipFile.getEntry( "plugin.xml" ) == null )
                 {
-                    return;
+                    return false;
+                }
+
+                // Protect against zip-slip (before applying any file-system modifications).
+                if ( JiveGlobals.getBooleanProperty( "plugins.loading.zipslipDetection.enabled", true ) )
+                {
+                    for ( Enumeration e = zipFile.entries(); e.hasMoreElements(); )
+                    {
+                        JarEntry entry = (JarEntry) e.nextElement();
+                        Path entryFile = dir.resolve( entry.getName() );
+                        if ( !entryFile.normalize().toAbsolutePath().startsWith( dir.normalize().toAbsolutePath() ) )
+                        {
+                            throw new RuntimeException( "Plugin contains content that is outside of target plugin directory (possible zipslip attack)" );
```

Comment:
would this exception tell the user which plugin had content outside the target?

Reply:
It does so indirectly. This exception is caught a few lines down, and logged as:

> An exception occurred while trying to extract plugin '{}'